src/plugins/footnotes.ts

```diff
@@ -0,0 +1,249 @@
+    inlineFootnotes: [],
+    unusedFootnoteIDs: new Set(),
+    footnotes: [],
+    labels: new Map(),
+  }
+
+  const footnotes = findChildrenByType(element[0], schema.nodes.footnote)
+  const footnoteIDs = new Set()
+  footnotes.forEach(({ node, pos }) => {
+    fn.footnotes.push([node as FootnoteNode, element[1] + pos + 1])
+    fn.unusedFootnoteIDs.add(node.attrs.id)
+    footnoteIDs.add(node.attrs.id)
+  })
+
+  let index = 0
+  const inlineFootnotes = findChildrenByType(
+    container[0],
+    schema.nodes.inline_footnote
+  )
+  inlineFootnotes.sort((a, b) => a.pos - b.pos)
+  const orderedFootnoteIDs: string[] = []
+  inlineFootnotes.forEach(({ node, pos }) => {
+    const inlineFootnote = node as InlineFootnoteNode
+    const rids = inlineFootnote.attrs.rids
+    if (rids.some((rid) => !footnoteIDs.has(rid))) {
+      return
+    }
+    if (container[1]) {
+      pos += container[1] + 1
+    }
+    fn.inlineFootnotes.push([node as InlineFootnoteNode, pos])
+    rids.forEach((rid) => {
+      if (orderedFootnoteIDs.indexOf(rid) < 0) {
+        const label = container[1]
+          ? String(++index)
+          : generateAlphaFootnoteLabel(index++)
+        fn.labels.set(rid, label)
+        fn.unusedFootnoteIDs.delete(rid)
+        orderedFootnoteIDs.push(rid)
+      }
+    })
+  })
+  fn.unusedFootnoteIDs.forEach((id) => orderedFootnoteIDs.push(id))
+  fn.footnotes.sort(
+    ([a], [b]) =>
+      orderedFootnoteIDs.indexOf(a.attrs.id) -
+      orderedFootnoteIDs.indexOf(b.attrs.id)
+  )
+
+  return fn
+}
+
+const getVersion = (
+  $new: FootnotesElementState,
+  $old?: VersionedFootnotesElementState
+) => {
+  const nids = $new.footnotes.map(([node]) => node.attrs.id)
+  const oids = $old?.footnotes.map(([node]) => node.attrs.id)
+  if (
+    isEqual(nids, oids) &&
+    isEqual($new.unusedFootnoteIDs, $old?.unusedFootnoteIDs)
+  ) {
+    return $old?.version
+  }
+  return String(version++)
+}
+
+export default (props: EditorProps) => {
+  return new Plugin<FootnotesPluginState>({
+    key: footnotesKey,
+    state: {
+      init(config, instance): FootnotesPluginState {
+        return buildPluginState(instance.doc)
+      },
+      apply(tr, value, oldState, newState): FootnotesPluginState {
+        const $old = footnotesKey.getState(oldState)
+        if (!tr.docChanged && $old) {
+          return $old
+        }
+        return buildPluginState(newState.doc, $old)
+      },
+    },
+    appendTransaction(transactions, oldState, newState) {
+      const $old = footnotesKey.getState(oldState)
+      const $new = footnotesKey.getState(newState) as FootnotesPluginState
+
+      const tr = newState.tr
+
+      $new.footnotesElements.forEach((newState, key) => {
+        const element = newState.element[0]
+        const pos = newState.element[1]
+        const oldState = $old?.footnotesElements.get(key)
+        const footnotes = newState.footnotes.map(([node]) => node)
+
+        if (newState.version !== oldState?.version) {
```

Comment:
the process of versioning seems to be very complicated, why are we versioning each footnote element? And why are we creating a new footnote element when the versions don't match?

Reply:
For each element, the version changes every time the order of the footnotes in that element change. The reason each element has its own version is so that we only re-render elements (and inline footnotes) that changed. When the versions for an element don't match, that means the order of footnotes changed, so we need to update the doc to reflect that order (`footnotes` in the state are ordered).

Reply:
Hm, wouldn't it be simpler to update decoration of the node from the plugin, so the updates on the nodes will be triggered organically without the need to wrap with mark and version?

Reply:
Not sure what you mean. The plugin indeed adds decorations to the nodes so that they're updated. If you're talking about replacing the element, that was there even before and I don't see a way around it.

Reply:
What I mean is instead of delegating to the nodeView the display of the label, just to make a decoration that will show the label so it will issue only meaningful updates. Same way as you set decoration 'footnote-selected' on line 297. Perhaps it may limit the amount of updates since if for some nodes the label didn't change, the decoration won't be updated. What do you think?

Reply:
A decoration will not be enough unfortunately as the footnote itself has to move. Otherwise, the footnote text will appear in the wrong order.